main/User_config.h

```diff
@@ -105,12 +109,19 @@ const byte mac[] = {0xDE, 0xED, 0xBA, 0xFE, 0x54, 0x95}; //W5100 ethernet shield
 #  define mqtt_topic_max_size  50
 #  define mqtt_max_packet_size 128
 #endif
-char mqtt_user[parameters_size] = "your_username"; // not compulsory only if your broker needs authentication
-char mqtt_pass[parameters_size] = "your_password"; // not compulsory only if your broker needs authentication
-char mqtt_server[parameters_size] = "192.168.1.17";
-char mqtt_port[6] = "1883";
-char mqtt_topic[mqtt_topic_max_size] = Base_Topic;
-char gateway_name[parameters_size * 2] = Gateway_Name;
+
+#ifndef MQTT_USER_DEFAULT
+#  define MQTT_USER_DEFAULT "your_username"
+#endif
+#ifndef MQTT_PASS_DEFAULT
+#  define MQTT_PASS_DEFAULT "your_password"
+#endif
+#ifndef MQTT_MQTT_DEFAULT
+#  define MQTT_MQTT_DEFAULT "192.168.1.17"
+#endif
+#ifndef MQTT_PORT_DEFAULT
+#  define MQTT_PORT_DEFAULT "1883"
+#endif
```

Comment:
I would not name these _DEFAULT

Reply:
Any naming recommendation? As C language is case sensitive, I could use names like `MQTT_USER` without the `_DEFAULT`, but thought it might be confusing with the corresponding `char[]` created in `main.ino` with the lowercase name `mqtt_user`. 

Also, should each variable check for if it's already defined, or should it be grouped into two logical blocks? Meaning, should we require the user to set not set any values, or set them all.

```
if !defined(MQTT_USER_DEFAULT) || !defined(MQTT_PASS_DEFAULT)
#  define MQTT_USER_DEFAULT "your_username"
#  define MQTT_PASS_DEFAULT "your_password"
#endif

if !defined(MQTT_MQTT_DEFAULT) || !defined(MQTT_PORT_DEFAULT)
#  define MQTT_MQTT_DEFAULT "192.168.1.17"
#  define MQTT_PORT_DEFAULT "1883"
#endif
```

Reply:
The internal name `user_name` in the main.ino file is not visible by the user, the users should only look into the config. And for developer it should be clear that one is a macro and the other a variable. Also the macro is only used at one place, so only user_name should be used in the code.

I would not group the definitions, as you probably will not change the default mqtt port but the ip/hostname.

Btw. the `mqtt_server_name` should also be handled like the other variables, maybe we can merge it with the `mqtt_server` variable, as they are handled the same (I could currently enter the hostname in mqtt_server and it would work). @1technophile We also have to think about the string length of the parameters, as they might not be long enough. FYI, the password length of 20 is not enough for my generated passwords.

Reply:
I've removed the `_DEFAULT` postfix and renamed `MQTT_MQTT` to be `MQTT_SERVER` so it's more descriptive as to its purpose.

I agree that `mqtt_server_name` is redundant as the PubSubClient library handles DNS and IP addresses the same and both can be configured using the same config location.
I've removed the `mqtt_server_name` definition the User_config.h and the reduced the code block from `main.ino`. 

https://github.com/1technophile/OpenMQTTGateway/blob/40a69037e5066b00a0ecc9bc88e9b6a5c4b176ca/main/main.ino#L539-L545

Reply:
> I agree that mqtt_server_name is redundant as the PubSubClient library handles DNS and IP addresses the same and both can be configured using the same MQTT_SERVER setting.

This was already part of a other #696 PR, you should rebase your branch on the latest development branch, so there are no merge conflicts.

Reply:
Didn't see that PR. I guess I need to rebase more often. I didn't realize I was that far behind on my fork.

Why are we keeping `mqtt_server_name` in PR #696? Couldn't `mqtt_server_name` and `mqtt_server` be combined?

Reply:
`mqtt_server_name` and `mqtt_server` can be combined but the PR was only to fix TLS connections, so I didn't refactor it as much.

Reply:
> Didn't see that PR. I guess I need to rebase more often. I didn't realize I was that far behind on my fork.

@jmw6773 you may use `git rebase -i upstream/development` after you added the upstream repository and fetched it, to cleanup this pull request.

Reply:
@Legion2 , thank you for the command. I'm still learning how to use `git`. I've rebased to upstream/development, then to this branch.

Reply:
your rebase didn't work, now the commits are duplicated. After a rebase you should not do a `git pull`! because it will merge it with itself. After a rebase you must force push the change to the branch or you could push them to a new branch. I will fix that and push it to a new branch.

Reply:
here is how it looks after the rebase, all the merges are gone and I removed the "revert" commits and commits already in the development branch https://github.com/1technophile/OpenMQTTGateway/compare/Move-MQTT-Config-rebased

Reply:
The rebase you did has the `Var naming and efficiency change ` commit where I removed `mqtt_server_name` and combined it with `MQTT_SERVER`.

Should I try rebasing on my side again?

Reply:
> The rebase you did has the Var naming and efficiency change commit where I removed mqtt_server_name and combined it with MQTT_SERVER.

yes this was the plan

> Should I try rebasing on my side again?

do you know how git rebase -i works?

Reply:
Not really. I know what it's suppose to be used for, but not really how to use it.

If it would be easier, we can can close this PR and create new one from the branch you created. 
Otherwise, I can refork, re-apply the changes, and then open a new PR.
I've begun to lose confidence in fixing this PR.

Reply:
If you want to learn some cool git feature look into `git rebase -i`, it must be used with care but you can fix any git problem with it. On the other hand doing rebases there is much which can go wrong, so you have to really know what you are doing.

We can use the `Move-MQTT-Config-rebased` branch I created, it only contains the necessary commits of this PR. I can force push them on your branch then, we don't have to create a new PR. If I force push on your branch you should delete your local branch (`git checkout development && git branch -D Move-MQTT-Config && git pull && git checkout Move-MQTT-Config`) to prevent any new merge conflicts.

Reply:
I'll definitely look into it. I got far enough to know that I could cause problems. I'll create a test repo to practice with.

I don't know if you need it, but I've invited you as a Collaborator to my fork so you can push any changes you need. I'll backup my local clone and re-clone my fork locally once you've make any changes.

Sorry for making you do all of this extra work.

Reply:
thanks for the invitation, but it is not required, as there is a PR setting in GitHub that allows upstream collaborators to push to the branch of the pull request, even if the branch comes from a fork.

> I'll backup my local clone and re-clone my fork locally once you've make any changes.

It should be enough to do a `git fetch` and `git reset --hard` when you are on the Move-MQTT-Config branch.

> Sorry for making you do all of this extra work.

No problem

Reply:
*`git reset --hard origin/Move-MQTT-Config`

Reply:
I've hard reset on all my local copies.